codalab/lib/bundle_fuse.py

```diff
@@ -64,6 +65,65 @@ def yank_path(self, path):
                 if path in key[0]:
                     del self._caches[func][key]
 
+class ByteRangeReader(object):
+
+    def __init__(self, client, bundle_uuid, timeout=60, max_num_chunks=100, chunk_size=None):
+        if chunk_size is None:
+            chunk_size = 1 * 1000 * 1000 # 1 MB
+
+        self.chunk_size = chunk_size
+        self.max_num_chunks = max_num_chunks
+        self.timeout = timeout
+        self.client = client
+        self.bundle_uuid = bundle_uuid
+        self.cache = OrderedDict() # (path, chunk_id) -> (time, bytearray)
+
+    def read(self, path, length, offset):
+        start_offset = offset
+        end_offset = offset + length - 1
+
+        start_chunk = self._get_chunk_id(start_offset)
+        end_chunk = self._get_chunk_id(end_offset)
+        arr = ''
+        for chunk_id in range(start_chunk, end_chunk + 1):
+            arr += self._fetch_chunk(path, chunk_id)
+        return arr[offset - start_chunk * self.chunk_size:offset - start_chunk * self.chunk_size + length]
+
+    def _fetch_chunk(self, path, chunk_id):
+        '''
+        Fetch a chunk from the cache, or with the client as necessary.
+        Refreshes chunks that are older than self.timeout
+        Only save full chunks to the cache. Partial chunks are not cached.
```

Comment:
Say that partial chunks are only at the end of the file because they could grow...

Make the assumptions clear.